oldabe/money_in.py

```diff
@@ -177,26 +178,58 @@ def generate_transactions(amount, attributions, payment_file, commit_hash):
     return transactions
 
 
-def total_amount_paid(for_email):
+def get_existing_itemized_payments():
+    # TODO
+    itemized_payments = []
+    itemized_payments_file = os.path.join(ABE_ROOT, ITEMIZED_PAYMENTS_FILE)
+    try:
+        with open(itemized_payments_file) as f:
+            for (
+                email,
+                fee_amount,
+                project_amount,
+                attributable,
+                payment_file,
+            ) in csv.reader(f):
+                itemized_payment = ItemizedPayment(
+                    email,
+                    fee_amount,
+                    project_amount,
+                    attributable,
+                    payment_file,
+                )
+                itemized_payments.append(itemized_payment)
+    except FileNotFoundError:
+        itemized_payments = []
+    return itemized_payments
+
+
+def total_amount_paid_to_project(for_email, new_itemized_payments):
     """
-    Calculates the sum of a single user's attributable payments for
-    determining how much the user has invested in the project so far.
-    Non-attributable payments do not count towards investment.
+    Calculates the sum of a single user's attributable payments (minus
+    fees paid towards instruments) for determining how much the user
+    has invested in the project so far. Non-attributable payments do
+    not count towards investment.
     """
+    all_itemized_payments = (
+        get_existing_itemized_payments() + new_itemized_payments
+    )
     return sum(
-        p.amount
-        for p in get_all_payments()
+        p.project_amount
+        for p in all_itemized_payments
         if p.attributable and p.email == for_email
     )
 
 
-def calculate_incoming_investment(payment, price):
+def calculate_incoming_investment(payment, price, new_itemized_payments):
     """
     If the payment brings the aggregate amount paid by the payee
     above the price, then that excess is treated as investment.
     """
-    total_payments = total_amount_paid(payment.email)
-    previous_total = total_payments - payment.amount
+    total_payments = total_amount_paid_to_project(
+        payment.email, new_itemized_payments
```

Comment:
Going from all payments to only the new ones is a significant change. It introduces the assumption that previous payments you are no longer looking at were already accounted for.

Is there a way to make this work with all previous itemized payments?

Reply:
Good call out! So we actually do use all itemized payments when we tally the total amount paid. Inside `total_amount_paid_to_project`, we combine the sums of existing itemized payments that have already been recorded in the file and these `new_itemized_payments`.

The reason we have to pass in the new itemized payments is because we haven't actually recorded them yet since we wait until payments have been processed without errors before recording anything (new valuation, renormalized attributions, and now, itemized payments). If we switch to a sqlite database at some point, this issue will be solved. We will be able to write to the itemized payments table within a transaction and query the whole table to get the correct total instead of passing new itemized payments down the chain.